rfd/0119-aws-api-integration-using-oidc.md

```diff
@@ -0,0 +1,686 @@
+* IdP: Identity Provider is a system that manages identities.
+* JWT: JSON Web Token is a json encoded token that contains a set of claims signed by an entity.
+* JWKS: JSON Web Key Set is a set of public keys that IdP consumers must use to validate a signed JWT.
+
+## Why
+
+Our current discovery process for AWS resources requires multiple steps before the user gets to the "aha moment".
+
+For example, to set up an RDS DB the user must do the following:
+- enter database name
+- enter Database endpoint, port, resource id and AWS Account ID
+- start a Database Agent by running a shell script (`database-install.sh`) in a EC2 instance
+- create an IAM Policy
+- add Database User and Database Name to the current user's traits
+- user can connect to this Database
+
+Configuring the IAM Policy, installing the Database Agent or setting up the traits the is sometimes challenging and we lose users before they get the "aha moment".
+
+Using AWS OIDC Integration allows Teleport services to call AWS API methods to obtain most of this information instead of asking the user.
+## How
+
+AWS allows the [set up](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) of an IdP using OIDC providers.
+Each provider has a set of roles, which limit their access.
+
+Turning Teleport into an OIDC IdP, and asking the user to create a new OIDC IdP that consumes the Teleport OIDC endpoints, allows us to easily call AWS API endpoints to, for instance, list RDS instances and their details.
+
+When configuring the provider, we'll need an AWS Role which Teleport uses to issue API Calls.
+
+To store the configuration above, we'll create a new resource Kind: `Integration`.
+It'll leverage the `subkind` prop to distinguish future integrations.
+
+```proto
+// IntegrationV1 represents a connection between Teleport and some other 3rd party system.
+// This connection allows API access to that service from Teleport.
+// Each Integration instance must have a SubKind defined which identifies the external system.
+message IntegrationV1 {
+  // Header is the resource header.
+  ResourceHeader Header = 1;
+
+  // Spec is an Integration specification.
+  IntegrationSpecV1 Spec = 2 ;
+}
+
+// IntegrationSpecV1 contains properties of all the supported integrations.
+message IntegrationSpecV1 {
+  oneof SubKindSpec {
+    // AWSOIDC contains the specific fields to handle the AWS OIDC Integration subkind
+    AWSOIDCIntegrationSpecV1 AWSOIDC = 1;
+  }
+
+  // IntegrationStatus contains the current integration status.
+  enum IntegrationStatus {
+    INTEGRATION_STATUS_UNSPECIFIED = 0;
+
+    // Integration is not running because it was just created or was disabled by the user.
+    INTEGRATION_STATUS_PAUSED = 1;
+
+    // Integration is running.
+    INTEGRATION_STATUS_RUNNING = 2;
+
+    // Integration has an error and that should be fixed before receiving requests.
+    INTEGRATION_STATUS_ERROR = 3;
+  }
```

Comment:
The integration may become unhealthy (OIDC IdP was removed from AWS, Role doesn't allow for required api calls, thumbprint changed).
In that case, there will be a Cluster Alert and the Integration will switch to Error Status.

The Paused status is to allow the user to disable the Integration but not removing it.
Maybe that's something we can work later on, and is not really needed for this first iteration.

Reply:
If the proxy rotates its SSL certificate the integration will also become unhealthy as AWS saves and compares the fingerprint of the SSL certificate on setup

Reply:
thumbprint*

Reply:
It's not quite that drastic, it actually uses the fingerprint of the TLS certificate's CA. So you can rotate certificates as long as it's issued by the same CA.

Reply:
Ah I misunderstood. Cheers

Reply:
@marcoandredinis I understand the general intent but tbh it just raises more questions: how would Teleport detect that it's unhealthy and why, and whether it's a permanent or temporary unhealthiness, and would we detect it's healthy after being unhealthy, and what does the user need to do in that case? And so on.

We did have some ideas (or rather, "aspirations") for upstream resource health checks (e.g. https://github.com/gravitational/teleport/issues/20544) but it feels like it's a large enough problem that would require a separate RFD and consistent approach we can use for all resources.

I would cut this from the scope of this RFD and implementation for now.

Reply:
Sounds fair :+1: 
I'll remove the status from the RFD and implementation (backport to V12 is not merged yet, so it will wait to include this removal)